### ISSUE TITLE
No audio after 1st loop

### ISSUE THREAD

**Issue by bouncyball-git** [ilia3101/MLV-App#25]
While playing back the clip with loop (never ending playback) turned on audio plays only during 1st loop then there is a silence :) Is that expected behavior?

**Comment by masc4ii** [comment]
No, this is not expected. For me it (mostly) works. But sometimes strange things happen with audio. And I also know why... it is very bad implemented :-D 
I have plans how to make it better, but I need some time to do so! Thanks for remind me! ;-)

**Comment by masc4ii** [comment]
Okay... I was wrong. Does not work for me too. I'll debug it. Quick fix does not work yet :(

**Comment by masc4ii** [comment]
Good. Or not good. I have to reimplement it when I want to solve it :-(

**Comment by masc4ii** [comment]
Please try it out. Audio playback completely rewriten.

**Comment by bouncyball-git** [comment]
Now  audio works but is out of sync (delayed) after 1st loop. Sometimes it  catches up and gets in sync on some loops, very strange :), but most of  the time is delayed though. I'm talking about Linux here.
--

**Comment by masc4ii** [comment]
Hehe, okay, I'll look for that! Thx.

**Comment by masc4ii** [comment]
Yes, I think I found the bug. See latest commit. Hopefully the delay is better (or even away).

**Comment by bouncyball-git** [comment]
0c5ed50 this is the commit with latest change of 'video_mlv.c'

**Comment by bouncyball-git** [comment]
Maybe it's on Linux only but sync is off and now some loops are silent again. If i press the back to 1st frame button slowly it is working if I press it quickly then silence kicks in. Maybe just set some time out before starting the playback again and it sorts things out?

**Comment by masc4ii** [comment]
Hm... the "out-of-sync-problem" is strange. Ok - exactly correct it can nearly never be, because audio and video is realized independent. And the slower the frame rate, the bigger the sync delay.
But it should never be silent. That is not good. I added some lines, please try again.
And pressing 1st frame button was not implmeneted in reference to audio playback (did that now).
Thx again!

**Comment by masc4ii** [comment]
New idea: audio is now exactly synced when drawing the frame. Negative site: audio playback starts on first frame redraw, not when pressing play button. Hope that is okay. But this should also be good for slow computers. Checked in into repos.

**Comment by bouncyball-git** [comment]
Can not compile: [qrc_ressources.cpp] Error 1

**Comment by masc4ii** [comment]
Sry. Now you can.

**Comment by bouncyball-git** [comment]
+        <file>RetinaIMG/Actions-color-picker-grey-icon.png</file>
+        <file>RetinaIMG/Actions-color-picker-icon.png</file>

files missing

**Comment by bouncyball-git** [comment]
Yes compiled: but silence almost all the time :-/ on some loops enabled from th ebeginning 3-4 loops silence.

**Comment by masc4ii** [comment]
Okay... I can't understand that... :-( 
I think I need a Linux now.

**Comment by bouncyball-git** [comment]
Does it work on mac as expected?

**Comment by masc4ii** [comment]
Yapp. 100%. Can loop it 20x and there is always audio.

**Comment by bouncyball-git** [comment]
The bad thing is that now if I play without loop enabled there is no sound at all :(

**Comment by masc4ii** [comment]
Maybe (if you like) try to debug it. The AudioPlayback class is very easy. It can load and unload the audio, start, stop and jump to position (in dependency to the frame number). Unload is automized, but you could also do it manually.
So for playing audio: load, jump to position and then play should bring playback. You could also do lines like:
`qDebug() << "some text" << variableX << ... ;`
for easy debugging, and you see that in the Application output window of creator.

**Comment by bouncyball-git** [comment]
Thank you I'll try this today :)

**Comment by masc4ii** [comment]
Checked on Windows7Pro: no problem.

**Comment by bouncyball-git** [comment]
After commit 04e0ad640e4872a73336e49dd03a118bee980947 audio is not working on Linux.
Also m_pAudioPlayback->jumpToPos( ui->horizontalSliderPosition->value() ) never worked properly. No matter which commit I'd roll back.

```
QAudioDeviceInfo info(QAudioDeviceInfo::defaultOutputDevice());
    if (!info.isFormatSupported(format)) {
        qWarning() << "Default format not supported - trying to use nearest";
        format = info.nearestFormat(format);
    }
    foreach (const QAudioDeviceInfo &deviceInfo, QAudioDeviceInfo::availableDevices(QAudio::AudioOutput))
        qDebug() << "Device name: " << deviceInfo.deviceName();
```
Use this code to check supported devices and their names. Seems all is ok and QT uses pulseaudio plugin on linux.

I'm lost. Even when audio worked before this commit - actually it kinda worked, loosing sync, muting when jump to pos and muting for no reason randomly.

**Comment by masc4ii** [comment]
Really sorry. Is that a Qt bug on Linux? That works so good on Windows & OSX... In the end, I don't understand why anything worked before.

**Comment by bouncyball-git** [comment]
Yes I know you switched to m_tryToSyncAudio variable and sync is used only in drawFrame. But as I already said the audio never worked under Linux as expected. Now it's not working at all and this is unfortunately expected ;). There is one project on github which adds support of jackaudio to QT. Maybe I'll try it someday. JACK Audio is for pro audio low latency stuff.

**Comment by masc4ii** [comment]
That it did not work when you opened the bug was expected. I loaded and started audio when pressing "play" and stopped audio when pressing "stop". So it was expected, if you use a loop, that there would be no audio. 
That's why I decided to implement an audio class, which has audio loaded the whole time and can jump on any time to any position. I am sad that this does not work on Linux, what is strange, because the used Qt methods are the same.

**Comment by bouncyball-git** [comment]
I liked new audio class. It's implemented the way it should be. I think the big mess in audio field of Linux is the reason (many sound systems). However, if you'll have time just try it on Linux, maybe you're gonna find something what I missed.

**Comment by seescho** [comment]
@bouncyball-git 

Succeded to give it a short try.

Here in Linux openSUSE I cannon hear sound on mlv-app playback, too. MLV App playback sends a pulseoudia-stream, which is recognized by my system, but I cannot hear anyhing. Screenshot:

https://seescho.files.wordpress.com/2017/10/pulseaudio.jpg

Exporting to avi gives a avi-file with sound.

**Comment by masc4ii** [comment]
Thanks! Very interesting! Did you enable "Export Audio" in Preferences? If yes, this sounds like a problem with loading the audio from the mlv file. 
Tried exporting a avi with sound on OSX - no problem.

**Comment by seescho** [comment]
Export to avi is OK with sound. No problem. The problem is on playback. I will try to nail down the commit, which causes this issue.

**Comment by masc4ii** [comment]
Ah sry. I read "avi-file without sound". My fault. I know which commit brings the problem. I rewrote the whole audio handling. The used Qt methods are still the same as before, so this very strange for me.

**Comment by seescho** [comment]
It´s the commit "better audio video sync for playback", right?. That´s the first one without sound on playback for me.

**Comment by masc4ii** [comment]
Before you really get audio??? Okaaaay... I thought [this commit](https://github.com/ilia3101/MLV-App/commit/b2c1afd0cb9aedbc2b6c3ca6b9bcf51a1ab5439c) is the last working for Linux.

**Comment by seescho** [comment]
On commit "added wbPicker icons and action..." I have audio on playback.

**Comment by masc4ii** [comment]
The only logic difference I see is, I call one stop() to much, if audio is not running yet and should be started. Made a fix for that. I don't think it will help, but please try out.
https://github.com/ilia3101/MLV-App/commit/04e0ad640e4872a73336e49dd03a118bee980947
Old line 2328, there is no stop. In the new version, always stop is called first. Maybe linux counts how often I stop and play audio and then nothing comes out... hahahaha.

**Comment by masc4ii** [comment]
If does not help, please try to add this to the AudioPlayback::play() function:
> void QAudioOutput::resume()
> Resumes processing audio data after a suspend().
So, `m_pAudioOutput->resume()` maybe changes anything...

**Comment by seescho** [comment]
No success with the new commit. I will try your other tip later

**Comment by masc4ii** [comment]
I did another commit.

**Comment by seescho** [comment]
No audio on playback with the other commit :)
Now it begins to be complicated to me. Not shure, what I must check now. I set a stoppoint on line 98 in AudioPlayback .cpp. Look at my screenshot, maybe this help?
https://seescho.files.wordpress.com/2017/10/another-fix.jpg

**Comment by masc4ii** [comment]
Hm... all looks good to me. Very identical to my debug output.
At last line in play() I added `qDebug() << m_pAudioOutput->state();` It tells me `ActiveState` after pressing the play button. Same for you?

**Comment by seescho** [comment]
Yes, says "ActiveState"

**Comment by masc4ii** [comment]
A little summary:
- Audio is loaded
- Functions are called
- Linux audio settings shows the connection to Qt via Pulseaudio
- Audio module is in ActiveState
- revisions ([older than this](https://github.com/ilia3101/MLV-App/commit/04e0ad640e4872a73336e49dd03a118bee980947)), where the audio functions were called from actions were working (with bad sync)
- revisions ([this and later](https://github.com/ilia3101/MLV-App/commit/04e0ad640e4872a73336e49dd03a118bee980947)), where the audio functions were called from functions, which were called from the timer were not working
- Problems only on linux. No problems on Win&OSX.

**Comment by seescho** [comment]
MainWindow.cpp
I changed lines 376 ff like this for a test
```
{
        m_tryToSyncAudio = false;
        m_pAudioPlayback->stop();
        m_pAudioPlayback->jumpToPos( 0 );
        m_pAudioPlayback->play();
    }
```
With this change, audio playback is running.

**Comment by masc4ii** [comment]
What? You just changed `ui->horizontalSliderPosition->value()` to 0 and you hear audio?
Does the position calculation in jumpToPos not work on linux? Or the seek? But what is strange: when starting audio after loading, the slider value is 0. So where is the difference? I still don't understand.

**Comment by seescho** [comment]
Doesn´t compile:

```
m_tryToSyncAudio = false;
m_pAudioPlayback->stop();
m_pAudioPlayback->jumpToPos( ui->horizontalSliderPosition->value() );
//m_pAudioPlayback->jumpToPos( 0 );
qDebug() << frame;
m_pAudioPlayback->play();
```

Error:

`/home/edgar/MLV App/audiobug/MLV-App-master/platform/qt/MainWindow.cpp:381: Fehler: ‘frame’ was not declared in this scope`

**Comment by masc4ii** [comment]
No sry... add it inside the jumpToPos() function in file AudioPlayback.cpp, please. Maybe also the value of position and audiostreamsize would be interesting. So the function could look like that:
```
//Jump to frame
void AudioPlayback::jumpToPos( int frame )
{
    if( !doesMlvHaveAudio( m_pMlvObject ) ) return;

    qint64 position = 4 * (qint64)( frame * getMlvSampleRate( m_pMlvObject ) / getMlvFramerate( m_pMlvObject ) );
    m_pAudioStream->device()->seek( position );
    qDebug() << frame << position << m_pAudioStream->device()->size();
}
```

**Comment by bouncyball-git** [comment]
as I mentioned above jumpToPos() never worked for linux.

**Comment by bouncyball-git** [comment]
this is the debug output when loop is enabled:
When loop is going from the end to the start sometimes 2 lines printed sometimes one.

1 8008 422400
0 0 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
0 0 422400
1 8008 422400
0 0 422400
0 0 422400
1 8008 422400
0 0 422400
0 0 422400
0 0 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400
0 0 422400
1 8008 422400

**Comment by masc4ii** [comment]
The debug log looks good, no error so far. And also when you got this output there was no audio? So the very last line is the one with seek(). But why does it work, when we hard code a 0, and it does not work when the algorithm writes a calculated 0 on seek? I still don't understand...

**Comment by bouncyball-git** [comment]
device()->seek( position ) always returns true

**Comment by masc4ii** [comment]
m_pAudioStream->device()->pos() gives back the position in the audio data. Is it equal to the position in seek( position ) ?

**Comment by seescho** [comment]
I guess, this stuff is far beyond my skills (atm). Some time ago, I bought a thick nice book to understand that C++ stuff a bit. But I´m on page 100 from more than 1000 pages :)

I did some more hardcoding like this: 
`m_pAudioPlayback->jumpToPos( 3 );`
Here are the results of the german jury :)

```
0 0 1075200
1 7680 1075200
2 15360 1075200
3 23040 1075200
10 76800 1075200
```

If I hardcode frames to 0, than I have audio output on playback. No audio with the other frames.

**Comment by masc4ii** [comment]
Thx. All these numbers are correct I think. But it only works for 0. So that is also far beyond my skills. :-(
In principle it is not that complicated: we have the frame number to which we want to sync. We now the number of samples per second, number of audio channels, number of bits per sample and number of frames per second. This is the formula from frame to position. With seek we just jump to the right position in the audio data. Function play() starts the playback at this position. I think the problem must be deeper than we think atm. The most strange is, that it works for the hard coded 0.

**Comment by masc4ii** [comment]
I again compared the [revision without audioclass](https://github.com/ilia3101/MLV-App/commit/b2c1afd0cb9aedbc2b6c3ca6b9bcf51a1ab5439c) with the latest one. And I found another difference. @bouncyball-git : With this very old revision, audio was starting from each position? Or also only from zero?
I found out that in this old version I sended the QObject handle from the application to QAudioOutput, which handles the playback. This was lost by implementing the AudioPlayback class. Now I changed it and QObject is sended. Please try latest commit...

**Comment by bouncyball-git** [comment]
Hey guys some progress here. It seems on linux Audio Buffer should be smaller. On my PC the smallest buffer working without hiccups is 63536 (I set to 131072). I also redone some stuff in mainwindow.cpp and stopping and resuming is working too. No clean loop support yet. 

Another thing: I think the timer should be started only when play button is pressed. If you have other reasons to leve timer running all the time please tell me.

**Comment by masc4ii** [comment]
You get audio on linux now? What exactly did you change?

The timer draws the picture. If you stop it, you won't see a frame. If you draw the frame by changing a value, the GUI will be veeeery slow.

**Comment by seescho** [comment]
@masc4ii Thanks for the explanation of the function.

@all
What happens with latest commit?

Starting playback without loop: no sound
If I jump back to frame 0, when playback is running, I have sound in some rare cases.
Starting playback with loop:
First loop: no sound
the following loops: sound

**Comment by masc4ii** [comment]
Funny... or not... but... let's see something positive: there is sound! :-)
Could you please change line 61 in AudioPlayback.cpp to:
`m_pAudioOutput->setBufferSize( 131072 );`

**Comment by seescho** [comment]
Works! (And I´m going back to my stars and planets, which need no audio :) )

**Comment by masc4ii** [comment]
Yeeeehaaaaa! THANK YOU ALL!
@bouncyball-git : same for you?

**Comment by bouncyball-git** [comment]
@masc4ii: If you mean planets and starts - no ;)
Smaller buffer size is just one correct step to the right direction (32mbyte buffer was too much really, even 1mb is too much, because it distracts smooth slider operation, skipping/resuming) I need to rewrite some stuff to be sure there is nothing can be done further.

**Comment by masc4ii** [comment]
@bouncyball-git : did you try the latest revision? I not only changed the buffer size. I also changed how QAudioOutput is called. Yes, I also thought when implementing that this is too much... but it was working on Win & OSX so I no more thought about that. Does it work for you now with this QObject change?

**Comment by bouncyball-git** [comment]
I'll try and report back :)

**Comment by masc4ii** [comment]
With the small buffer I now have the problem, that it only works with bilinear demosaic. When caching or Amaze rendering, it stops very often is is never in sync. I now remember this was the reasons why I made the buffer huuuuuuge and I never had a problem after that.

**Comment by masc4ii** [comment]
@bouncyball-git : Did you try again if it works now for yout too (since I changed the class interface)? And which buffer size is best for you?

**Comment by bouncyball-git** [comment]
This turned out to be a looong thread :)

@masc4ii:
Hey man. I want you to help me out and test my rewritten playback code [HERE](https://github.com/bouncyball-git/MLV-App/commits/master).

Recently I dug into your code deeply to understand how your playback engine works. After some deep investigation I came to conclusion that if I want to get smoother audio experience under Linux I must do something with frame dropping part. The reason was I wanted to decouple jumpSlider movement from real frame rendering.

Now in my modified version slider moves linearly alongside with audio playback and frame dropping does not affect slider movement at all. I got what I wanted but... there is a caveat. It works flawlessly if there is a need to drop no more than every other frame (e.g. if desired fps is 24 and achieved not less then 12 it works very good) If frame drawing takes more time than this there is a lag (e.g. amaze case and very slow dualiso processing case)

To completely deal with this lag, frame rendering should be started as a parallel thread of the timer event handler (timeEvent) and render exact required frame in the background. So I was reeealy lazy to learn how to do this and gave up at least for now.

Additionally I fixed all playback handlers to work properly in play and stopped state. Also if audio playback is checked in the menu and code detects lag it automatically turns on frame dropping and checks item in the menu. I must say I'm not 100% satisfied but smooth audio playback and overall smooth lagless playback is now WAY better under Linux.

I had no time to test it under win at the moment but will do this ASAP.

Please test it on Mac and I really need your opinion regarding these changes :-D

Thanks in advance
bb

**Comment by bouncyball-git** [comment]
@masc4ii: The answer to your previous question:
The 'this' class passing did nothing. e.g. it worked as it worked without it as well.
The best buffer for Linux is 128k. 512kb introduces lag or muting.
I #ifdef-ed the buffer size in my new version.

**Comment by masc4ii** [comment]
So it changed nothing? Okay... the loop problem is also still there? :-/

Wow... what a change... I only did a very quick test until now. I must look in detail what you did to respond better. 
But my quick test will make you very sad: on OSX there is absolutely no sync anymore now. If I press start, video & audio starts. When the audio is ready, video is approx. at the half, or a bit less. Video in this clip was 5fps before, and now is also 5fps. Anything with the fps-counter is also wrong: it jumps 23..5..23..5.. (but for 23 it is way to slow).
So on OSX I think it is not the effect you wanted to have... :-(
But I will look in the code when I have some more time. I am sure you had many good ideas! ;-)

**Comment by bouncyball-git** [comment]
> on OSX there is absolutely no sync anymore now. If I press start, video & audio starts. When the audio is ready, video is approx. at the half, or a bit less.

Haha! Funny thing is this was exactly as you say under Linux before I changed the code ;)
Please try to change the buffer to 128k on the mac and see if it changed anything ok?

**Comment by bouncyball-git** [comment]
//qDebug() << timeDiff << frameDuration + 1; activate this line and see the timings. I guess I have far more powerfull machine than you. Sadly... It's not gonna work out the way I did it. Well... there is no way to get is right without backgroung rendering task.

**Comment by masc4ii** [comment]
128k does not change anything. Audio is clear. But video is way to slow... far away from realtime (as I can name that dropmode realtime). I tried escho's version on Linux (via live CD) and this was working good for me. But this was the old 0.8 I think, where Loop & audio did not work.

The qDebug brings "40 41" very often ;-)

My machine is a 7years old Core2Duo 2.4GHz with 16GB RAM and SSD.

**Comment by bouncyball-git** [comment]
During playback first number should not be more than 80 (2 * second number), otherwise you are in a trouble ;) and the behavior you explained is the case.

**Comment by bouncyball-git** [comment]
I could not run drawFrame as a thread. The error was the timer can not be started from the other thread...

**Comment by masc4ii** [comment]
Ah okay... first time it is "0 41", "116 41", "40 41",...
Packing drawFrame in another thread is not that easy as under C (or I don't know how).

**Comment by masc4ii** [comment]
Why do you like to run drawFrame in another thread? I think the only advantage would be a more responsible UI. The sync won't be better.

**Comment by bouncyball-git** [comment]
> "536 41", "77 41", "275 41", "47 41"

Whoo... :)

> Why do you like to run drawFrame in another thread?

2 parallel processe:
1. timerEven with correct fps and slider incremented by one
2. background rendering of a frame with calculated index (1, 5, 10 etc...) to have smooth but dropped frame rate.

Otherwise timerEvent can not be invoked by the timer every needed time until routine exits. Now timer generates events but timerEvent can not be run on every event unfortunately (can not be run 2 parallel timerEvents at the same time). 

That is why 'm_frameStillDrawing' flag is always true 

```
void MainWindow::drawFrame( void )
{
    m_frameStillDrawing = true;
    ...
    m_frameStillDrawing = true;
}
```

and checking it in the place of 'Trigger Drawing' is pointless (as it is right now):

```
//Trigger Drawing
if( !m_frameStillDrawing && m_frameChanged && !m_dontDraw && !m_inOpeningProcess )
{
      m_frameChanged = false; //first do this, if there are changes between rendering
      drawFrame();
      //Allow interaction while playback
      //qApp->processEvents();
      ...
}
```
Actually It would be complicated to sync those 2 processes and I guess your smart jumpSlider skipping is the way to go with slow rendering If audio is ok :)

**Comment by masc4ii** [comment]
If just audio is too slow for you... the Audio can easily be brought to another thread, because it is a QObject now. To try it out:

Add in MainWindow.h, line 139:
`QThread *m_pAudioThread;`

And in MainWindow.cpp, line 65:
```
m_pAudioThread = new QThread;
m_pAudioPlayback->moveToThread( m_pAudioThread );
```

Maybe the picture processing blocks the audio engine, if everything runs in one thread.

**Comment by bouncyball-git** [comment]
@masc4ii 

I reworked the frame dropping code and as I mentioned earlier implemented the separate thread for the frame rendering using QThread subclass. 

It's a proof of my concept only and needs further discussion, testing and there are stuff to finish.

It works good even in the most demanding case (amaze + dualiso) and adapts to the CPU speed itself.
It is here, in my [repo](https://github.com/bouncyball-git/MLV-App/commits/master)

Please compile and try on your slow mac and do not pay attention to FPS on the status bar it does not show the correct numbers yet.

Eagerly awaiting for your reply :P
bb

**Comment by masc4ii** [comment]
@bouncyball-git : 
I also wrote a render thread now, you can test it here: https://github.com/masc4ii/MLV-App
Just the render routine is a extra thread now, all other stuff works like before. So the fps should be identical. But the sliders are very responsive all the time. And maybe audio is working for...

**Comment by bouncyball-git** [comment]
Hey Man THX! 

You brought my concept and code to the whole new level with scrupulousness and finesse specific to you :-D and I appreciate this very much!

I've been playing with this version whole morning today and wanna share some thoughts.
The good:
1. Under Linux audio playback now starts as expected after pressing play button.
2. The responsiveness and CPU usage of the app overall is very good. 
Not so good:
1. Audio playback stops somewhat later after video stops.
2. I'm experiencing random crashes (on random places of the getMlvProcessedFrame8()) when switching video in the session list during playback. It crashes despite correctly handling "isIdle" function. Something simultaneously in the background accesses the (valid/freed) data when quickly changing videos in the session (those crashes are common in my version too).
3. Here is the [video](http://nic.caucasus.net/mlvapp/?dir=&download=Garbage_In_the_buffer.mp4) demonstrating supposedly uninitialized/old image buffer access before crash.

**Comment by masc4ii** [comment]
+1&2: Cool! Fine that audio works! Yes, a lot of inspriration came from your code ;-)
-1: When you press stop, or when the video is on its end? The second is normal, because the audio data we get is longer than the video data. Ilia wanted to fix that using timestamps.
-2&3: thx for the video. With that I can work! ;-) But on none of my Macs that happens. 

For me the only negativ was, that the render thread stopped working exactly one time. All was working, but the frame was black until app restart. But I could not reproduce...

**Comment by bouncyball-git** [comment]
Yes audio runs after video reaches the end. I'll test this further.

Yeah, black stalled frame with audio and slider running happened to me with my version as well. It was the case when I tortured the session list a lot :)

**Comment by masc4ii** [comment]
There was access on the frame after openMlv deleted the structure. I changed some lines. Please test ;-)
The black stalled frame was for me while playback... in one frame it stopped forever :o)

**Comment by bouncyball-git** [comment]
Ah yes, that one... 

Yesterday when I experiencing the issue I already moved the waiting loop before freeing the whole object (sorry that I forgot to tell you about this bug) but it changed nothing regarding garbage in the buffer. After that I reported this to you.

**Comment by bouncyball-git** [comment]
Now when you added qApp->processEvents() to the loop it is very interesting to test again. I'll check and report back.

**Comment by bouncyball-git** [comment]
Well that's it! I did not experienced the buffer with the garbage any more! :-D

Just one thing

issue: 
when playing one MLV, switching to the second one from session list causes audio to stay and play if second MLV does not have audio at all.

caused by these lines in the AudioPlayiback.cpp functions:
```
//No audio? Quit!
if( !doesMlvHaveAudio( m_pMlvObject ) ) return;
```

The Fix:

```
    //Waiting for frame ready because it works with m_pMlvObject
    while( m_frameStillDrawing ) {qApp->processEvents();}

+   //Unload audio
+   if( doesMlvHaveAudio( m_pMlvObject ) ) m_pAudioPlayback->unloadAudio();

    /* Destroy it just for simplicity... and make a new one */
    freeMlvObject( m_pMlvObject );
```

in openMlv() function

**Comment by bouncyball-git** [comment]
Overall this is the great progress with audio under Linux and I would love to merge those changes to master branch! :)

**Comment by bouncyball-git** [comment]
BUT... there is some issues with caching turned ON. When caching on and playback is active, switching videos in the session list causes total freese and then crash:

```
$ ./mlvapp
QObject::killTimer(): Error: timer id 6 is not valid for object 0x7ffc64ad34b0 (MainWindow, MainWindow), timer has not been killed
Segmentation fault (core dumped)
```

Unfortunately I don't have the time to track this issue right now.

**Comment by bouncyball-git** [comment]
I think second 1sec timer (m_timerCacheId) also should be killed and restarted when new MLV is loading.

**Comment by bouncyball-git** [comment]
Another crash log:

```
./mlvapp
m_timerId = 6, m_timerCacheId = 5  
QObject::killTimer(): Error: timer id 4 is not valid for object 0x7ffe63eb0e60 (MainWindow, MainWindow), timer has not been killed
[xcb] Unknown sequence number while appending request
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
mlvapp: ../../src/xcb_io.c:161: append_pending_request: Assertion `!xcb_xlib_unknown_seq_number' failed.
Aborted (core dumped)

```

**Comment by masc4ii** [comment]
I changed some things. You can try again! ;-)

**Comment by bouncyball-git** [comment]
Tried and everything works!!! :-D

Now... I have one simple and obvious fix for not playing (long) audio after video is stopped.

```
void MainWindow::playbackHandling(int timeDiff)
{
    if( ui->actionPlay->isChecked() )
    {
        //when on last frame
        if( ui->horizontalSliderPosition->value() >= ui->horizontalSliderPosition->maximum() )
        {
+           m_pAudioPlayback->stop();
            if( ui->actionLoop->isChecked() )
            {
                //Loop, goto first frame
                ui->horizontalSliderPosition->setValue( 0 );
```

**Comment by bouncyball-git** [comment]
If on mac and windows everything is OK then I'm happy to announce that audio issue on Linux resolved. 

We DID IT and this thread well worth of its 92 comments :)

There is a caching + rendering thread issue remains though, what are your thoughts about it?

**Comment by bouncyball-git** [comment]
BTW the problem when caching freezes whole app during switching session videos with active playback happens in main branch too.

**Comment by bouncyball-git** [comment]
With windows I'll check tomorrow. 

I discovered that if you maximize the window during playback video stops but audio continues to play :)

**Comment by masc4ii** [comment]
Try the latest commit in my repos. ;)

**Comment by bouncyball-git** [comment]
Yes now it works! If we could solve the caching issue then we can even close this thread :). Or... merge changes, close this thread and open new issue.

**Comment by masc4ii** [comment]
I'll merge this evening. Do you add a new issue for the caching issue? On OSX I never had this problem, so I can't write a good description of the problem.

**Comment by bouncyball-git** [comment]
> On Windows 32bit I have one last problem now: sometimes (not always) the app freezes when pressing
> play or stop for up to 5sec (only when AudioOutput is enabled). Do you have that for Win64 too?
Yes the same ugly thing occurs with 64bit build

**Comment by bouncyball-git** [comment]
Ive got the proposition. Let's make audio buffer optional, e.g. adjustable by user through the File/Preferences.

**Comment by masc4ii** [comment]
Okay... but I am not sure if that solves the problem. I can adjust whatever... it feels always ugly atm. I also would do a 3rd ifdef for windows. But I don't find a completely working buffer value. With the very first audio solution the app was never frozen.

**Comment by masc4ii** [comment]
Tried again on another veeeery slow Win7 i5: audio starts without delay. When stopping a clip the 1st time, the app freezes 3sec, after that it does not freeze until you switch the clip and stop the new clip for the 1st time. Not nice... but the windows users will survive... they know they have a slow system :-D

**Comment by bouncyball-git** [comment]
> after that it does not freeze until you switch the clip and stop the new clip for the 1st time

Exactly, 1st time huge pause and then less pauses but always some small pause persists...

**Comment by bouncyball-git** [comment]
BTW why you used in your audio stopping/starting implementation resume alongside start and suspend alongside stop?

**Comment by masc4ii** [comment]
Without suspend audio runs until its end on OSX :-)

**Comment by bouncyball-git** [comment]
Hm... very strange behavior, on all 3 platforms QT audio has its quirks :)

**Comment by masc4ii** [comment]
Very unusual for Qt5 indeed. But I don't know a better audio library :-(

**Comment by masc4ii** [comment]
There is still this QMediaPlayer class. There is a setMedia, play, stop, setPosition - all we need. But I don't find a solution howto load data from a buffer. That only works with QUrl, so we need the wav file. But I don't like writing a file just to hear audio.

**Comment by masc4ii** [comment]
What minimal buffersize is working for you on Win? 32768 seems to work on my ultra slow i5 Win7. With that stop latency is very low.

**Comment by bouncyball-git** [comment]
If buffer is optional in the settings dialog the experimenting will be very easy.

**Comment by masc4ii** [comment]
I tried it on another Windows, and it was also working well. I don't know if I like the idea with the settings dialog, because I think such parameters are unneccessary for the user - "simple things" (in the eyes of the user) like audio playback must just work...

**Comment by bouncyball-git** [comment]
Yes I understand but we have issues with this buffer all the time and I think there should be some sort of fix without compiling. We set default values for all OSes and then the option will be there grayed out. We can not say 100% that someone somewhere will not have some issues with the audio any more. The fix then will be to play with this buffer.

**Comment by bouncyball-git** [comment]
Can we close this thread?

**Comment by masc4ii** [comment]
No one cried about having still problems... so I would like to close! Closed! :o)
